### PR TITLE
Reduce direct memory OOM chances on broker with a per server query response size budget

### PR DESCRIPTION
**Pull request by vvivekiyer**
When server responds for a query with a large response, the broker can potentially crash with direct memory OOM. 

In PR https://github.com/apache/pinot/pull/11496 - a fix was added to restart the Netty Channel in such scenarios. This will result in all active queries failing with an error. This is a good fix to have to protect Brokers in this extreme case. 

However, to reduce the probability of such events happening and contain the impact for other queries, this PR introduces a threshold at the Server. If the serialized query response at a Server exceeds the threshold, the query is failed.  The overall threshold for a query is set using a config. This budget is divided across all Servers processing the query. 

cc: @siddharthteotia @jasperjiaguo  @gortiz @ege-st @dinoocch

### CODE REVIEW

File: pinot-core/src/main/java/org/apache/pinot/core/query/scheduler/QueryScheduler.java
Comment (gortiz) [inline pinot-core/src/main/java/org/apache/pinot/core/query/scheduler/QueryScheduler.java:178]: Can we add this before merge?

File: pinot-core/src/main/java/org/apache/pinot/core/query/scheduler/QueryScheduler.java
Comment (gortiz) [inline pinot-core/src/main/java/org/apache/pinot/core/query/scheduler/QueryScheduler.java:180]: I think we should add a specific QueryException for this case

File: pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/OfflineClusterIntegrationTest.java
Comment (gortiz) [inline pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/OfflineClusterIntegrationTest.java:612]: I don't think it is a good idea to assert on messages. We should instead create a specific QueryException with its own error code and assert that the exception error code is that one.

**Comment by gortiz** [comment]
I've added some notes, but I think the PR is in good shape

**Comment by ege-st** [comment]
Cool, thanks! I think reducing the likelyhood of DM OOMs is the logical next step.

How will a user determine what the threshold should be? A critical requirement for a feature like this is to introduce it such that it does suddenly start breaking queries that were working before the threshold was enabled; so, a guide on how to determine what the threshold should be is, I think, critical for safe adoption.

**Comment by vvivekiyer** [comment]
> How will a user determine what the threshold should be? 

@ege-st 
Good question. We did think about this. The ideal answer is that the value is use case dependent.
However, a safe starting point is to set the threshold limit to max direct memory size  - this value is conservative to not break existing queries and at the same time, it reduces the probability of brokering OOMing. For the usecase that we saw in our production environment, this value works. 
Also note that this setting is disabled by default in this PR. It can be enabled on a case-by-case basis.

File: pinot-spi/src/main/java/org/apache/pinot/spi/utils/CommonConstants.java
Comment (siddharthteotia) [inline pinot-spi/src/main/java/org/apache/pinot/spi/utils/CommonConstants.java:371]: The limit that is exposed to the user  as a config is at the query level and we internally compute the budget per server. Correct ?

In that case, why should this per server limit be put in the CommonConstants file ?

File: pinot-spi/src/main/java/org/apache/pinot/spi/utils/CommonConstants.java
Comment (siddharthteotia) [inline pinot-spi/src/main/java/org/apache/pinot/spi/utils/CommonConstants.java:322]: What does -1 mean ?

File: pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java
Comment (siddharthteotia) [inline pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java:1682]: Are we introducing this as a query option or instance config or both ?

File: pinot-core/src/main/java/org/apache/pinot/core/query/scheduler/QueryScheduler.java
Comment (jasperjiaguo) [reply pinot-core/src/main/java/org/apache/pinot/core/query/scheduler/QueryScheduler.java:180]: or use QUERY_CANCELLATION_ERROR for which we've set the error code to 503(Service Unavailable)

**Review by jasperjiaguo — COMMENTED** [review]
LGTM over all, I think the behavior of having both broker config and query option to set this is 
 pretty much like the global threshold + query option we do for timeout, easpecially when the direct memory threshold can be conservative and not considering data skew. My only concern is the user need to be instructed to not abuse the option and kill the cluster, or we can have a swich to turn it off centrally (for security).

File: pinot-core/src/main/java/org/apache/pinot/core/query/scheduler/QueryScheduler.java
Comment (siddharthteotia) [inline pinot-core/src/main/java/org/apache/pinot/core/query/scheduler/QueryScheduler.java:174]: Is it better to do this sooner as the serialized response is being built ? There is a method that converts DataTable into toBytes(). The method does incrementally by going over various data structures in the DataTable. Ideally we can intersperse this check within that method at various points ?

On a related note, is it possible to do this check as segment level individual result is being combined by the CombineOperator ?

File: pinot-spi/src/main/java/org/apache/pinot/spi/utils/CommonConstants.java
Comment (ege-st) [reply pinot-spi/src/main/java/org/apache/pinot/spi/utils/CommonConstants.java:322]: I believe -1 means that there is no limit, but I think it would be more clear and lead to simpler code if we make the default `Long.MAX_VALUE`.

File: pinot-core/src/main/java/org/apache/pinot/core/query/scheduler/QueryScheduler.java
Comment (ege-st) [reply pinot-core/src/main/java/org/apache/pinot/core/query/scheduler/QueryScheduler.java:174]: +1. The sooner will kill it the less time and memory we spend on it.

**Review by ege-st — CHANGES_REQUESTED** [review]

File: pinot-spi/src/main/java/org/apache/pinot/spi/utils/CommonConstants.java
Comment (Jackie-Jiang) [inline pinot-spi/src/main/java/org/apache/pinot/spi/utils/CommonConstants.java:321]:
```suggestion
    public static final String CONFIG_OF_MAX_SERVER_RESPONSE_SIZE_BYTES = "pinot.broker.max.server.response.size.bytes";
```

File: pinot-spi/src/main/java/org/apache/pinot/spi/utils/CommonConstants.java
Comment (Jackie-Jiang) [inline pinot-spi/src/main/java/org/apache/pinot/spi/utils/CommonConstants.java:371]:
```suggestion
        public static final String MAX_SERVER_RESPONSE_SIZE_BYTES = "maxServerResponseSizeBytes";
```

File: pinot-spi/src/main/java/org/apache/pinot/spi/utils/CommonConstants.java
Comment (Jackie-Jiang) [inline pinot-spi/src/main/java/org/apache/pinot/spi/utils/CommonConstants.java:322]: I'd suggest putting a default (e.g. 1GB if not too big)

**Review by Jackie-Jiang — COMMENTED** [review]
You may also consider adding a table level override for this config

File: pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java
Comment (vvivekiyer) [reply pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java]: Since this is a instance config that doesn't change unless restarted, I've actually added a log to indicate this value. I can add a gauge metric for sure, but do you think it'll be helpful?

File: pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java
Comment (vvivekiyer) [reply pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java]: There are 2 configs involved here:

There's a broker instance config that defines what the maximum response size for single query can be. If this is set, this total value is divided by the number of routed servers to determine the `per_server_budget`. This `per_server_budge` is set as a query config very similar to how we handle timeout today (the alternative is to make thrift changes to pass this info).

File: pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java
Comment (vvivekiyer) [reply pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java]: This is not possible today as we check for zero servers in L651. 
Nevertheless, I added a safeguard condition.

File: pinot-core/src/main/java/org/apache/pinot/core/query/scheduler/QueryScheduler.java
Comment (vvivekiyer) [reply pinot-core/src/main/java/org/apache/pinot/core/query/scheduler/QueryScheduler.java]: I did evaluate that option. But we currently do not have a good mechanism for handling exceptions during serializing and effectively return to the user. I've added a TODO here.

File: pinot-core/src/main/java/org/apache/pinot/core/query/scheduler/QueryScheduler.java
Comment (vvivekiyer) [reply pinot-core/src/main/java/org/apache/pinot/core/query/scheduler/QueryScheduler.java]: Good point. Added.

File: pinot-core/src/main/java/org/apache/pinot/core/query/scheduler/QueryScheduler.java
Comment (vvivekiyer) [reply pinot-core/src/main/java/org/apache/pinot/core/query/scheduler/QueryScheduler.java]: Good idea, done.

File: pinot-spi/src/main/java/org/apache/pinot/spi/utils/CommonConstants.java
Comment (vvivekiyer) [reply pinot-spi/src/main/java/org/apache/pinot/spi/utils/CommonConstants.java:321]: This is the max response size per **"query"**. So I believe the existing name makes more sense. Wdyt?

File: pinot-spi/src/main/java/org/apache/pinot/spi/utils/CommonConstants.java
Comment (vvivekiyer) [reply pinot-spi/src/main/java/org/apache/pinot/spi/utils/CommonConstants.java]: Set to Long.MAX_VALUE

Comment (vvivekiyer) [reply pinot-spi/src/main/java/org/apache/pinot/spi/utils/CommonConstants.java]: Added Long.MAX_VALUE as default

File: pinot-spi/src/main/java/org/apache/pinot/spi/utils/CommonConstants.java
Comment (vvivekiyer) [reply pinot-spi/src/main/java/org/apache/pinot/spi/utils/CommonConstants.java]: That is correct. I've answered the reasoning in another comment above.

There's a broker instance config that defines what the maximum response size for single query can be. If this is set, this total value is divided by the number of routed servers to determine the per_server_budget. This per_server_budge is set as a query config very similar to how we handle timeout today (the alternative is to make thrift changes to pass this info).

File: pinot-spi/src/main/java/org/apache/pinot/spi/utils/CommonConstants.java
Comment (ege-st) [reply pinot-spi/src/main/java/org/apache/pinot/spi/utils/CommonConstants.java:321]: I would agree that `query` makes more sense.

File: pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java
Comment (gortiz) [reply pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java]: It can be useful to be used as a max. For example we use to start with Xmx = Xms, but we use the max heap value to calculate the % of heap used. We could do something like that with this metric.

**Comment by siddharthteotia** [comment]
I am still not convinced of how we have wired the config aspect.

At a high level, I suggest we do the same thing as `timeOut` config since that is exactly what we seem to be doing here with an instance level config and then a query level config (in the form of query option) IIUC. So may be should also leverage queryConfig at the table level as well and have the preference as:

`queryOption (query level) > table level > instance level `

Also this may be a nit but the following 2 configs in CommonConstants is not super intuitive to me.

`public static final long DEFAULT_MAX_QUERY_RESPONSE_SIZE_BYTES = Long.MAX_VALUE;`

`public static final String MAX_SERVER_RESPONSE_SIZE_BYTES = "maxServerResponseSizeBytes";`

One problem I see is that while (1) is an instance config but tells for the query and therefore we internally compute the per server threshold / budget during routing depending on how many servers we are routing to.

However, (2) seems to be the queryOption way of doing things but here we are having the user directly specify the server level budget instead of overall. Why this difference ?

File: pinot-spi/src/main/java/org/apache/pinot/spi/utils/CommonConstants.java
Comment (vvivekiyer) [reply pinot-spi/src/main/java/org/apache/pinot/spi/utils/CommonConstants.java:321]: Added tableConfig and override sequence. Thanks for the suggestion.

The reasoning behind adding a broker level instance config was - the broker ultimately should decide how much response size it should get for each query (depending on it's direct memory limits). If this broker instance config is set, we use that to set the query option to limit response size.

File: pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java
Comment (vvivekiyer) [reply pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java]: Added guage.

**Comment by vvivekiyer** [comment]
@siddharthteotia Added tableConfig and overriding sequence.

The reasoning behind adding a broker level instance config was - the broker ultimately should decide how much response size it should get for each query (depending on it's direct memory limits). If this broker instance config is set, we use that to set the query option to limit response size.

I'm open to suggestions if the broker instance config doesn't make sense.

**Comment by Jackie-Jiang** [comment]
I understand the intention of tracking broker side total memory usage, but evenly splitting it into server side memory limit might cause inconsistent behavior because the actually limit relies on the fanout, and it assumes evenly distribution of the results from each server.
Since the direct control is size limit per server, we should at least allow configuring it directly. If it is not configured, we can fall back to broker total memory limit.

**Comment by vvivekiyer** [comment]
@Jackie-Jiang got it. Based on the suggestions, I've added a server side limit in the table config. Would you prefer that we add an instance config for the server in addition to the table config?

**Comment by Jackie-Jiang** [comment]
@vvivekiyer I'm thinking 2 configs - `perServer` & `perQuery`. Both of them can be set in 3 places: query option > table config > broker instance config (per server one can also be configured on broker). `perServer` takes precedence, and if not configured, fall back to `perQuery`. So overall the fallback order should be `perServer` query option > `perQuery` query option > `perServer` table config > `perQuery` table config > `perServer` instance config > `perQuery` instance config

We need some documentation to clear document how these 2 configs are used

**Comment by vvivekiyer** [comment]
@Jackie-Jiang  added perServer and perQuery configs. Please take a look.

File: pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java
Comment (Jackie-Jiang) [inline pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java:1697]: (minor) Put this in a local variable to avoid reading it twice

File: pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java
Comment (Jackie-Jiang) [inline pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java:688]: (minor)
```suggestion
        numServers += offlineRoutingTable.size();
```

File: pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java
Comment (Jackie-Jiang) [inline pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java:691]: (minor)
```suggestion
        numServers += realtimeRoutingTable.size();
```

File: pinot-common/src/main/java/org/apache/pinot/common/metrics/BrokerMeter.java
Comment (Jackie-Jiang) [inline pinot-common/src/main/java/org/apache/pinot/common/metrics/BrokerMeter.java:103]: I don't see much value for this metric

**Review by Jackie-Jiang — APPROVED** [review]
LGTM otherwise

File: pinot-common/src/main/java/org/apache/pinot/common/metrics/BrokerMeter.java
Comment (vvivekiyer) [reply pinot-common/src/main/java/org/apache/pinot/common/metrics/BrokerMeter.java]: Removed this metric.

**Review by siddharthteotia — APPROVED** [review]

**Comment by Jackie-Jiang** [comment]
@vvivekiyer Have you added these configs to the pinot doc?

**Pull request merged**
